File: source/downloads/code/LeetCode/Java/Subsets.java
public class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        List<Integer> tmp = new ArrayList<Integer>();

        Arrays.sort(nums);
        dfs(ret, tmp, nums, 0);

        return ret;  
    }


    public void dfs(List<List<Integer>> ret, List<Integer> tmp, int[] nums, int index){
        
        ret.add(new ArrayList<Integer>(tmp));
        
        for(int i=index; i < nums.length; i++){
            tmp.add(nums[i]);
            dfs(ret, tmp, nums, i+1);
            tmp.remove(tmp.size()-1);
        }

    }
}

File: downloads/code/LeetCode/Python/Subsets.py
class Solution(object):
    def find_sets(self,nums,ret,tmp,index,n):
        if len(tmp) == n:
            ret.append(tmp[:])
            return
        for i in range(index,len(nums)):
            tmp.append(nums[i])
            self.find_sets(nums,ret,tmp,i+1,n)
            tmp.pop()
        return 

    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ret = []
        if not nums:
            return ret
        ret.append([])
        nums = sorted(nums)
        for i in range(1,len(nums)+1):
            self.find_sets(nums,ret,[],0,i)
        return ret
